File: .leetcode/47.全排列-ii.cpp
/*
 * @lc app=leetcode.cn id=47 lang=cpp
 *
 * [47] 全排列 II
 */
#include<bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<int>>ans;
    vector<int>temp;
    void backtracking(vector<int>&nums,vector<int>&n,vector<int>freq){
       if(temp.size() == nums.size()){
           ans.push_back(temp);
           return;
       }
       for(auto i : n){
           if(freq[i + 10]){
               freq[i + 10]--;
               temp.push_back(i);
               backtracking(nums,n,freq);
               temp.pop_back();
               freq[i + 10]++;
           }
       }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<int>freq(21,0);
        vector<int>n;
        for(auto i : nums){
            if(freq[i + 10] == 0)n.push_back(i);
            freq[i + 10]++;
        }
        backtracking(nums,n,freq);
        return ans;
    }
};
// @lc code=end

File: .leetcode/47.全排列-ii.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
 * @lc app=leetcode.cn id=47 lang=java
 *
 * [47] 全排列 II
 */

// @lc code=start
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    List<Integer> temp = new ArrayList<>();
    void dfs(int n,int[] nums,boolean[] visited){
        if(temp.size() == n){
            ans.add(new ArrayList<>(temp));
            return;
        }
        for(int i = 0;i < n;i++){
            if(i > 0 && nums[i] == nums[i - 1] && !visited[i - 1]){
                continue;
            }
            if(visited[i] == false){
                visited[i] = true;
                temp.add(nums[i]);
                dfs(n,nums,visited);
                visited[i] = false;
                temp.remove(temp.size() - 1);
            }
        } 
    }
    public List<List<Integer>> permuteUnique(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        boolean[] visited = new boolean[n];
        dfs(n, nums, visited);
        return ans;
    }
}
// @lc code=end

